### PR TITLE
RSDK-9060 Allow custom first run timeout configuration

### PR DESCRIPTION
**Pull request by maximpertsov**
### Description

Add a user-supplied `first_run_timeout` parameter to the module config.

Secondarily, add a custom Duration field that can be JSON (un)marshaled

### Prerequisites

* [x] https://github.com/viamrobotics/goutils/pull/381

### Manual Testing

* https://github.com/viamrobotics/rdk/pull/4511#issuecomment-2448535770
* Ran Fuzz on custom Duration test for 5 min without detecting a failure: `go test go.viam.com/rdk/utils -fuzz=FuzzDurationJSONRoundtrip -fuzztime=5m`

### TODO

* [x] Test custom `Duration` field
* [x] ~Move custom [Duration] to goutils - see [this comment](https://github.com/viamrobotics/rdk/pull/4511#pullrequestreview-2409161831) for details.~ Done: https://github.com/viamrobotics/goutils/pull/381
* [x] Add unit test for config serialization with `first_run_timeout`

### CODE REVIEW

File: module/modmanager/manager_test.go
Comment (maximpertsov) [inline module/modmanager/manager_test.go:1]: Note for reviewers - this file is much easier to review with [white-space changes omitted](https://github.com/viamrobotics/rdk/pull/4511/files?diff=split&w=1).

File: module/modmanager/manager_test.go
Comment (maximpertsov) [inline module/modmanager/manager_test.go:1582]: I wonder how portable this assertion is

**Comment by cheukt** [comment]
have you tested with a local/cached config?

**Comment by maximpertsov** [comment]
> have you tested with a local/cached config?

testing now - will report results shortly

**Comment by maximpertsov** [comment]
>     * there is a parsing bug - duration isn't getting parsed properly

fixed with https://github.com/viamrobotics/rdk/pull/4511/commits/fe5e3d426f8ec6b169b7dac00782fb880bad8e29. This lets me test with a local config. I get the expected output when I set a custom timeout:

config:

```json
{
  "components": [
    {
      "name": "mysensor",
      "namespace": "rdk",
      "type": "sensor",
      "model": "viam:sensor:linux-wifi",
      "attributes": {}
    }
  ],
  "modules": [
    {
      "first_run_timeout": "1ms",
      "type": "registry",
      "name": "5784b2a8-d84f-473b-9be5-9a2abe7e8752_my_setup_phase",
      "module_id": "5784b2a8-d84f-473b-9be5-9a2abe7e8752:my_setup_phase",
      "version": "0.0.12",
      "executable_path": "REDACTED",
      "first_run": "REDACTED"
    }
  ]
}
```

output:
```
2024-10-30T22:01:29.179Z        INFO    rdk.modmanager.first_run        modmanager/manager.go:1099      executing first run script      {"module":"5784b2a8-d84f-473b-9be5-9a2abe7e8752_my_setup_phase","path":"/home/mp/.viam/packages/data/module/5784b2a8-d84f-473b-9be5-9a2abe7e8752-my_setup_phase-0_0_12-linux-amd64/first_run.sh"}
2024-10-30T22:01:29.181Z        ERROR   rdk.modmanager.first_run        modmanager/manager.go:1176      first run script failed {"error":"signal: killed","module":"5784b2a8-d84f-473b-9be5-9a2abe7e8752_my_setup_phase","path":"/home/mp/.viam/packages/data/module/5784b2a8-d84f-473b-9be5-9a2abe7e8752-my_setup_phase-0_0_12-linux-amd64/first_run.sh"}
2024-10-30T22:01:29.181Z        ERROR   rdk     impl/local_robot.go:1263        error executing setup phase     {"module":"5784b2a8-d84f-473b-9be5-9a2abe7e8752_my_setup_phase","error":"signal: killed"}
```

**Comment by maximpertsov** [comment]
> * this field might not be flowing from app properly

As for this one, I guess it's blocked on [this ticket](https://viam.atlassian.net/browse/RSDK-8879).

File: utils/duration.go
Comment (maximpertsov) [inline utils/duration.go:38]: this is [copy-pasta from stackoverflow](https://stackoverflow.com/questions/48050945/how-to-unmarshal-json-into-durations) - I will simplify to only handle only strings.

File: utils/duration.go
Comment (maximpertsov) [inline utils/duration.go:1]: this file needs a test obviously, but let me know anyone disagrees with adding a custom `Duration` field for parsing purposes. we definitely have other fields in RDK that could benefit from this.

File: utils/duration.go
Comment (benjirewis) [inline utils/duration.go:11]: That's a pretty unfortunate oversight from Golang. So we need JSON (un)marshaling in addition to proto (un)marshaling because we want to specify durations in JSON? Is this the first JSON duration field we've ever had? I guess the other "duration" settings we have are through environment variables (`VIAM_RESOURCE_CONFIGURATION_TIMEOUT` + `VIAM_MODULE_STARTUP_TIMEOUT`). I'm assuming we don't want to use an env variable for `first_run_timeout`?

File: utils/duration_test.go
Comment (benjirewis) [inline utils/duration_test.go:12]: Can you explain what the fuzz test is doing here? Generating similar test cases to the ones you're `Add`ing? I imagine most fuzz tests result in error?

File: module/modmanager/manager_test.go
Comment (benjirewis) [inline module/modmanager/manager_test.go:1570]: Is 1ms supposed to be small enough to not let first run succeed? Might as well do a nanosecond in that case.

File: module/modmanager/manager_test.go
Comment (benjirewis) [reply module/modmanager/manager_test.go:1585]: Seems fine to me for now; thanks for comment. Doubt the assertion will hold on windows but it might on mac.

File: utils/duration.go
Comment (maximpertsov) [reply utils/duration.go:11]: this is still WIP, so TBD on whether this ends up being ultimate solution. we handle other config duration fields (e.g. session heartbeat window) by maintaining a separate struct that has a "string" field for each duration field. this intermediate struct is validated and at that point we convert said strings to [time.Duration]s.

File: utils/duration_test.go
Comment (maximpertsov) [reply utils/duration_test.go:12]: still WIP, but my goal is to demonstrate that for every JSON input - roundtrip (un)marshalling to my custom duration datatype gives the same result as using time.ParseDuration and back.

as I'm doing this I think my whole methodology might be flawed so may backtrack all of this.

Comment (maximpertsov) [reply utils/duration_test.go]: updated the test with comments on my methodology and I think it's sound now

File: utils/testdata/fuzz/FuzzDurationJSONRoundtrip/771e938e4458e983
Comment (maximpertsov) [inline utils/testdata/fuzz/FuzzDurationJSONRoundtrip/771e938e4458e983:1]: this generated input caused an earlier implementation of Duration marshal/unmarshal to fail the fuzz test. this input will be tested against moving forward.

File: module/modmanager/manager_test.go
Comment (maximpertsov) [inline module/modmanager/manager_test.go:1438]: this test already existed, I just decided to pull it out into a subtest

File: config/module.go
Comment (maximpertsov) [inline config/module.go:44]: `time.Duration` fields cannot be JSON unmarshaled. the pattern we've used to get around this is creating a parallel struct where all `time.Duration` fields are strings. this parallel struct's custom unmarshal method then parsed these fields using `time.ParseDuration`. this works, but requires us to always maintain a parallel struct. I think a better approach is to have a custom duration field that supports (un)marshaling, which is what I implement in this PR.

File: utils/duration_test.go
Comment (maximpertsov) [inline utils/duration_test.go:13]: just to be clear, no special command are required to run this test - all seed inputs (i.e. the `f.Add(input)` lines) will be tested along with any generated input ([example](https://github.com/viamrobotics/rdk/pull/4511/files#diff-d58ec87018d1433c0d2d9c5d8cafb91df7df5c0b11e159d216db471f322daa73)).

new generated inputs might be created by running `go test go.viam.com/rdk/utils -fuzz=FuzzDurationJSONRoundtrip -fuzztime=5m` (the timeout is optional). only inputs that result in a failing test actually result in a new generated input AFAICT - the idea being that failures represent regressions that should be tested against moving forward.

File: module/modmanager/manager_test.go
Comment (maximpertsov) [reply module/modmanager/manager_test.go]: woah, the test actual fails if the timeout is a nanosecond! the assertion that fails expects the script to have a non-zero exit code. this flies in the face of the testing schema I describe [here](https://github.com/viamrobotics/rdk/pull/4511#discussion_r1823372603). instead of checking for process errors, maybe it's most portable to check for context deadline exceeded error.

Comment (maximpertsov) [reply module/modmanager/manager_test.go]: :laughing: amazing, so asserting on a context deadline exceed error (https://github.com/viamrobotics/rdk/pull/4511/commits/6d2e8cfc78556248e2115bacd1a89331d943b932) passes for a nanosecond timeout but fails for a millisecond timeout...

i'm actually not sure what the best approach is anymore... my current mental model says that if the timeout is extremely fast, the process doesn't have enough time to even start so there is no process error but will be context deadline error. however, if the timeout is slow enough to allow the process to start but fast enough to pass before the process finishes, then we will get a process error.

maybe we test for both?

Comment (maximpertsov) [reply module/modmanager/manager_test.go]: > maybe we test for both?

ended up going just that: https://github.com/viamrobotics/rdk/pull/4511/commits/6fea0fd274b270ea55f4e2b2c65216ddac623186

will be monitoring for flakes but so far this seems to be consistent locally

File: utils/duration_test.go
Comment (maximpertsov) [reply utils/duration_test.go]: lemme know if this explanation suffices: https://github.com/viamrobotics/rdk/pull/4511/files#r1825152416

File: utils/testdata/fuzz/FuzzDurationJSONRoundtrip/771e938e4458e983
Comment (cheukt) [inline utils/testdata/fuzz/FuzzDurationJSONRoundtrip/771e938e4458e983:1]: should we delete this file then?

**Review by cheukt — COMMENTED** [review]
I think I found what app was missing, we need to update the code around https://github.com/viamrobotics/app/blob/71188c3d002dc9e11b67be6934246723ca1cd339/robotservice/config_retriever.go#L776 to include the first run timeout field. In this case, do you think it'd be better to keep the utils.Duration field in RDK or in goutils?

File: utils/testdata/fuzz/FuzzDurationJSONRoundtrip/771e938e4458e983
Comment (maximpertsov) [reply utils/testdata/fuzz/FuzzDurationJSONRoundtrip/771e938e4458e983:1]: so this scenario will continue to be tested, so I think it's fine to leave in as is. see https://github.com/viamrobotics/rdk/pull/4511/files#r1825152416 for more details

File: utils/duration.go
Comment (maximpertsov) [reply utils/duration.go:11]: i ended up successfully implementing the custom duration that supports marshaling/unmarshaling - i think using this new field is the approach i would prefer

**Comment by maximpertsov** [comment]
> In this case, do you think it'd be better to keep the utils.Duration field in RDK or in goutils?

good find. I suppose it should be in goutils then

File: go.mod
Comment (maximpertsov) [inline go.mod:448]: TODO: revert before merging

**Comment by maximpertsov** [comment]
> > In this case, do you think it'd be better to keep the utils.Duration field in RDK or in goutils?
> 
> good find. I suppose it should be in goutils then

done! https://github.com/viamrobotics/goutils/pull/381

**Comment by maximpertsov** [comment]
> > ```
> > * there is a parsing bug - duration isn't getting parsed properly
> > ```
> 
> fixed with [a39c4ae](https://github.com/viamrobotics/rdk/commit/a39c4ae5116b934ab157f9a5a0796c03efc0fbe8). This lets me test with a local config. I get the expected output when I set a custom timeout:
> 
> config:
> 
> ```json
> {
>   "components": [
>     {
>       "name": "mysensor",
>       "namespace": "rdk",
>       "type": "sensor",
>       "model": "viam:sensor:linux-wifi",
>       "attributes": {}
>     }
>   ],
>   "modules": [
>     {
>       "first_run_timeout": "1ms",
>       "type": "registry",
>       "name": "5784b2a8-d84f-473b-9be5-9a2abe7e8752_my_setup_phase",
>       "module_id": "5784b2a8-d84f-473b-9be5-9a2abe7e8752:my_setup_phase",
>       "version": "0.0.12",
>       "executable_path": "REDACTED",
>       "first_run": "REDACTED"
>     }
>   ]
> }
> ```
> 
> output:
> 
> ```
> 2024-10-30T22:01:29.179Z        INFO    rdk.modmanager.first_run        modmanager/manager.go:1099      executing first run script      {"module":"5784b2a8-d84f-473b-9be5-9a2abe7e8752_my_setup_phase","path":"/home/mp/.viam/packages/data/module/5784b2a8-d84f-473b-9be5-9a2abe7e8752-my_setup_phase-0_0_12-linux-amd64/first_run.sh"}
> 2024-10-30T22:01:29.181Z        ERROR   rdk.modmanager.first_run        modmanager/manager.go:1176      first run script failed {"error":"signal: killed","module":"5784b2a8-d84f-473b-9be5-9a2abe7e8752_my_setup_phase","path":"/home/mp/.viam/packages/data/module/5784b2a8-d84f-473b-9be5-9a2abe7e8752-my_setup_phase-0_0_12-linux-amd64/first_run.sh"}
> 2024-10-30T22:01:29.181Z        ERROR   rdk     impl/local_robot.go:1263        error executing setup phase     {"module":"5784b2a8-d84f-473b-9be5-9a2abe7e8752_my_setup_phase","error":"signal: killed"}
> ```

I tested again after importing `Duration` from goutils and the results were the same

File: config/config_test.go
Comment (maximpertsov) [inline config/config_test.go:1224]: repurposed to test a variety of marshall/unmarshal paths

File: config/config_test.go
Comment (maximpertsov) [inline config/config_test.go:1228]: i kept an `expected` field to match the original test, but so far it seems like we expect a `config.Config -> JSON -> config.Config` round trip to result in the same loaded config, so we could consider removing this at some point. I know `UnmarshalJSON` makes adjustments in some cases, but maybe we should architect away from this pattern.

File: go.mod
Comment (benjirewis) [inline go.mod:87]: What's up with this weird pseudoversion? Maybe try `GOPROXY=direct go get -u go.viam.com/utils`?

File: config/module.go
Comment (benjirewis) [inline config/module.go:45]: What does a timeout of `0s` mean? Looks like you treat it as "use default" and not "infinite"? Seems fine to me, but maybe document here.

**Review by benjirewis — APPROVED** [review]
LGTM; just a question or two.

File: config/module.go
Comment (maximpertsov) [reply config/module.go:45]: oh yeah that's a good catch, will document this explicitly for sure

Comment (maximpertsov) [reply config/module.go:45]: @benjirewis is a timeout of `0s` actually infinite or immediate? my understanding is that it's immediate. this doesn't change your suggestion for documentation - I'm still going to do that - but just want to confirm my understanding.

Comment (maximpertsov) [reply config/module.go:45]: confirmed for myself, seems like `0s` or less is immediate [per my testing](https://goplay.tools/snippet/AMF5jPG2sxc).

Comment (maximpertsov) [reply config/module.go:49]: let me know if this doc works for you: https://github.com/viamrobotics/rdk/pull/4511/commits/03b6aa2fa23041d71e3de5492a61dc1baf239ee3

File: go.mod
Comment (maximpertsov) [reply go.mod]: I upgraded goutils to `HEAD` without releasing a new version initially, but now there is a proper version out so this is fixed: https://github.com/viamrobotics/rdk/pull/4511/commits/f9d7211f14f8206c268ff97902178a46879c25dd

File: config/module.go
Comment (benjirewis) [inline config/module.go:45]:
```suggestion

```
[nit] Remove extra commented line (make it a newline).

File: config/module.go
Comment (benjirewis) [inline config/module.go:46]:
```suggestion
	// Setting this field to a zero value (e.g. 0 * [time.Second]) will 
	// set the first run timeout to the default value of 1 hour, which
```
Lines seem off here.

File: config/module.go
Comment (benjirewis) [reply config/module.go:49]: Seems good to me! Thanks.

> is a timeout of 0s actually infinite or immediate?

I guess in a formal sense: a timeout of 0s should put a deadline on the context of `time.Now()`, which would mean that attempting to use the context would almost certainly result in a `context deadline exceeded` error, as you would not be at `time.Now() + epsilon` where `epsilon` is some non-zero positive duration.

**Review by benjirewis — APPROVED** [review]
Still LGTM!

File: config/module.go
Comment (maximpertsov) [reply config/module.go:46]: sorry, bad formatting - good catch

**Review by cheukt — APPROVED** [review]

**Pull request merged**
